### PR TITLE
Introduce enum llama_ftype

### PR DESCRIPTION
**Pull request by sw**
As discussed in #678, this introduces `enum llama_ftype` to llama.h which represents the various file types used by llama.cpp

The goal is to improve maintainability of the code by not having integer literals in various places. I hope I didn't miss one.

I used the names and comments from @comex's pending changes to the Python scripts in #545. I have not touched the scripts so as not to cause conflicts with their work.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -100,7 +102,7 @@ struct llama_hparams {
     int32_t n_head  = 32;
     int32_t n_layer = 32;
     int32_t n_rot   = 64;
-    int32_t f16     = 1;
+    int32_t f16     = LLAMA_FTYPE_MOSTLY_F16;
```
Comment (prusnak) [inline llama.cpp:105]: How hard is it to change name of this field? This is a private struct, so should be easy, no?

**Comment by comex** [comment]
Thumbs-up from me on adding names.  But I’d argue that the file type field (`f16` in the file header) and tensor type field (`ftype` in the tensor header) should be two different enums.  For example, file type 3 represents what I called ‘mostly’ Q4_1, but tensor type 3 means that that particular tensor is (entirely) Q4_1.  A file with file type 3 has a mix of tensor types 0 (F32) and 3 (Q4_1).  I made this dissonance somewhat worse by adding file type 4, which has no tensor type equivalent; those files have a mix of tensor types 0 (F32), 1 (F16), and 3 (Q4_1).  

In the long run I’d like to see llama.cpp scrap the file type field entirely in favor of just using the tensor type and size fields to initialize the ggml tensor metadata.

**Comment by comex** [comment]
> We might want to make `enum ggml_type` match this, assuming that one's not written to the file. Which it isn't AFAICT.

It's not written to the file, so that might be a good idea.  No strong opinions.

**Review by ggerganov — APPROVED** [review]

**Comment by comex** [comment]
Even after my change makes the core loading path not care about file type, there will still be some uses, like in the quantization code (which has to accept a file type as an argument and then write it to the file header).  So a change like this would still be useful, and I’ll deal with any merge conflicts.  However, I still wouldn’t recommend merging this as-is, for the reason I mentioned before.

**Review by ggerganov — APPROVED** [review]

**Pull request merged**